src/rest/connection.cpp

```diff
@@ -0,0 +1,297 @@
+    if (this->mErrorFlag)
+    {
+        this->ErrorFormResponse(res);
+        this->SentResponse(res);
+        this->SetEndFlag(1);
+    }
+    else
+    {
+        if (this->mCallbackFlag)
+        {
+            duration = duration_cast<microseconds>(steady_clock::now() - this->mDiagInfo->mStartTime).count();
+            if (duration > 4 * sTimeout)
+            {
+                this->CallbackFormResponse(res);
+                this->SentResponse(res);
+                this->SetEndFlag(1);
+            }
+        }
+    }
+}
+
+void Connection::NonBlockRead()
+{
+    int received = 0;
+
+    while ((received = read(this->mFd, this->mReadBuf.get(), this->mBufLength)) > 0)
+    {
+        this->mReadContent += std::string(this->mReadBuf.get(), received);
+        this->mReadLength += received;
+    }
+    auto err = errno;
+    if (received == -1 && err != EAGAIN && err != EWOULDBLOCK)
+    {
+        this->mErrorFlag = 1;
+        this->mErrorCode = "error while read";
+    }
+}
+
+void Connection::CreateRequest(Request &aRequest)
+{
+    mParser->data = &aRequest;
+    http_parser_execute(mParser.get(), &mSettings, this->mReadContent.c_str(), this->mReadLength);
+}
+
+void Connection::SetEndFlag(int aFlag)
+{
+    this->mEndFlag = aFlag;
+    close(this->mFd);
+}
+bool Connection::CheckDiag(std::string &aRloc)
+{
+    if (!this->mCallbackFlag)
+        return false;
+    auto it = this->mDiagInfo->mNodeSet.find(aRloc);
+    if (it == this->mDiagInfo->mNodeSet.end())
+        return false;
+    else
+        return true;
+}
+
+void Connection::AddDiag(std::string aRloc, std::string aDiag)
+{
+    this->mDiagInfo->mDiagContent.push_back(aDiag);
+    this->mDiagInfo->mNodeSet.insert(aRloc);
+}
+
+void Connection::SetCallbackFlag(int aFlag)
+{
+    this->mCallbackFlag = aFlag;
+}
+void Connection::SetErrorFlag(int aFlag)
+{
+    this->mEndFlag = aFlag;
+}
+
+void Connection::ResetDiagInfo()
+{
+    this->mDiagInfo.reset();
+    mDiagInfo = std::unique_ptr<DiagInfo>(new DiagInfo(steady_clock::now()));
+}
+
+void Connection::ErrorFormResponse(Response &aResponse)
+{
+    aResponse.SetBody(this->mErrorCode);
+}
+
+void Connection::CallbackFormResponse(Response &aResponse)
+{
+    std::string str = mJsonFormater.VectorToJson(this->mDiagInfo->mDiagContent);
+    aResponse.SetBody(str);
+}
+
+void Connection::SetErrorCode(std::string aErrorCode)
+{
+    this->mErrorCode = aErrorCode;
+}
+
+requestHandler Connection::GetHandler(Request &aRequest)
+{
+    std::string url     = aRequest.getPath();
+    auto        Handler = this->mHander.GetHandler(url);
+    return Handler;
+}
+
+steady_clock::time_point Connection::GetConnectionStartTime()
+{
+    return this->mStartTime;
+}
+
+steady_clock::time_point Connection::GetDiagStartTime()
+{
+    return this->mDiagInfo->mStartTime;
+}
+
+void Connection::SentResponse(Response &aResponse)
+{
+    std::string data = aResponse.SerializeResponse();
+    write(this->mFd, data.c_str(), data.size());
```

Comment:
Can `write` fail?

Reply:
and we might not write all data, only part of it.